### DISCUSSION TITLE
Nested entity propagation eventhough it already exists

### DISCUSSION THREAD

**Discussion by reramjiawan** [mikro-orm/mikro-orm#1198] [Q&A] [answered]
**Describe the bug**
Nested entity should not be inserted into the database if it already exists.

**Stack trace**
```
  console.log
    [query] select `e0`.* from `b` as `e0` where `e0`.`value` = 'foo' limit 1 [took 1 ms]

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] begin

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] insert into `b` (`value`) values ('foo') [took 1 ms]

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] commit

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] select `e0`.* from `a` as `e0` left join `b` as `e1` on `e0`.`b_id` = `e1`.`id` where `e1`.`value` = 'foo' limit 1 [took 1 ms]

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] begin

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] insert into `b` (`value`) values ('foo') [took 1 ms]

      at Logger.log (packages/core/src/utils/Logger.js:22:14)

  console.log
    [query] rollback

      at Logger.log (packages/core/src/utils/Logger.js:22:14)


UniqueConstraintViolationException: insert into `b` (`value`) values ('foo') - SQLITE_CONSTRAINT: UNIQUE constraint failed: b.value

    at SqliteExceptionConverter.convertException (C:\Users\ravin\Desktop\mikro-orm\packages\sqlite\src\SqliteExceptionConverter.js:20:20)
    at SqliteDriver.convertException (C:\Users\ravin\Desktop\mikro-orm\packages\core\src\drivers\DatabaseDriver.js:171:54)
    at C:\Users\ravin\Desktop\mikro-orm\packages\core\src\drivers\DatabaseDriver.js:175:24
    at SqliteDriver.nativeInsert (C:\Users\ravin\Desktop\mikro-orm\packages\knex\src\AbstractSqlDriver.js:144:21)
    at ChangeSetPersister.persistNewEntity (C:\Users\ravin\Desktop\mikro-orm\packages\core\src\unit-of-work\ChangeSetPersister.js:55:21)
    at ChangeSetPersister.executeInserts (C:\Users\ravin\Desktop\mikro-orm\packages\core\src\unit-of-work\ChangeSetPersister.js:24:13)
    at UnitOfWork.commitCreateChangeSets (C:\Users\ravin\Desktop\mikro-orm\packages\core\src\unit-of-work\UnitOfWork.js:472:9)
    at UnitOfWork.persistToDatabase (C:\Users\ravin\Desktop\mikro-orm\packages\core\src\unit-of-work\UnitOfWork.js:434:13)

previous Error: insert into `b` (`value`) values ('foo') - SQLITE_CONSTRAINT: UNIQUE constraint failed: b.value

```

**To Reproduce**
```
import { Entity, ManyToOne, MikroORM, PrimaryKey, Property } from '@mikro-orm/core';
import { AbstractSqlDriver } from '@mikro-orm/knex';
import { SqlHighlighter } from '@mikro-orm/sql-highlighter';
import { classToPlain } from 'class-transformer';

@Entity()
export class B {

  @PrimaryKey()
  id!: number;

  @Property({ unique: true })
  value!: string;

}

@Entity()
export class A {

  @PrimaryKey()
  id!: number;

  @ManyToOne({ entity: 'B', eager: true })
  b!: B;

}

describe('GH issue 1192', () => {

  let orm: MikroORM<AbstractSqlDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [A, B],
      dbName: ':memory:',
      type: 'sqlite',
      debug: ['query', 'query-params'],
      highlighter: new SqlHighlighter(),
    });
    await orm.getSchemaGenerator().createSchema();
  });

  afterAll(() => orm.close(true));

  test('Nested entity should not get propagated if it already exists', async () => {

    // Uncomment this block to see the issue
    // const b = new B();
    // b.value = 'foo';

    // Uncomment this block to see the issue
    // await orm.em.findOne(b.constructor.name, classToPlain(b)) || await orm.em.persistAndFlush(b);

    const b2 = new B();
    b2.value = 'foo';

    const a = new A();
    a.b = b2;

    await orm.em.findOne(a.constructor.name, classToPlain(a)) || await orm.em.persistAndFlush(a);

    const b3 = new B();
    b3.value = 'foo';

    const a2 = new A();
    a2.b = b3;

    await orm.em.findOne(a2.constructor.name, classToPlain(a2)) || await orm.em.persistAndFlush(a2);

  });

});

```

**Expected behavior**
I expect that the existing entity does not get inserted into the database.

**Versions**

| Dependency | Version |
| - | - |
| node | 14.15.1 |
| typescript | 4.0.5 |
| mikro-orm | 4.3.4-dev.20 |
| mikro-orm/sqlite | 4.3.4-dev.20 |

**Comment by B4nan** [comment]
What are you expecting from this? What is that bit with class-transformer doing?

In general you can't use `class-transfomer` to work with MikroORM entities, it does not understand references and collections. You should do `wrap(a).toObject()` or `wrap(a).toJSON()` instead.

edit: but the problematic part is usually creating entity instances, using it for serialization might be possible (idk, never tried, as for me it is not needed at all)

**Comment by reramjiawan** [comment]
When you run this test without uncommenting the part that I mentioned the queries that are executed run fine and the existing entity B with value foo does not get propagated twice. When you uncomment the code I mentioned then suddenly it will try to propagate entity B twice even tho there should be a match in the database. The class transformer is just for the where clause.

**Comment by reramjiawan** [comment]
So when examining the query:
```select `e0`.* from `a` as `e0` left join `b` as `e1` on `e0`.`b_id` = `e1`.`id` where `e1`.`value` = 'foo'``` it should not propagate twice if value foo is already there and can be found but that seems not to be the case.

**Comment by B4nan** [comment]
Not really, the issue is in your code. It is not creating the same entity twice - you are. 

```ts
@Entity()
export class A {

  @PrimaryKey()
  id!: number;

  @ManyToOne({ entity: 'B', eager: true })
  b: B;

  constructor(body: { name: string }) {
    this.b = new B({ value: body.name }); // each new A will have new B with the same value
  }

}
```

**Comment by reramjiawan** [comment]
@B4nan Why did you close this seems you missed the point there clearly is a bug? I have updated my test case to make it less confusing.

**Comment by reramjiawan** [comment]
If you run the test case without uncommenting the code block I mentioned the queries that get executed are working as expected the first A entity propagates all changes and the second A entity doesn't need to propagate entity B because foo already exists. But as soon as you create B entity before the two A entities suddenly it doesnt work anymore.

**Comment by B4nan** [comment]
No, I did not miss the point, it sounds more like you did not understand my last message. Your code is creating multiple entities with `value: 'foo'`, that is your issue, not the ORM issuing multiple inserts for the same entity - you have multiple of those.

**Comment by reramjiawan** [comment]
That still doesn't explain why it does work as expected when I persist the two A entities without persisting the B entity before it. Those are both new B entities. The query that gets generated is 
```select `e0`.* from `a` as `e0` left join `b` as `e1` on `e0`.`b_id` = `e1`.`id` where `e1`.`value` = 'foo'``` which means there should be a match the second time around so there is no need for a persist for the second A entity. Which means my code is working as expected. ```await orm.em.findOne(a2.constructor.name, classToPlain(a2)) || await orm.em.persistAndFlush(a2);```. So that begs the question why it breaks when entity B is created before the two A entities.

**Comment by reramjiawan** [comment]
I realized my own mistake my apologies. The two a entities obviously work because the second time around the first a entity gets returned if that wasn't the case it would insert the new b again. Closed.

**Comment by B4nan** [comment, marked as answer]
> which means there should be a match the second time around so there is no need for a persist for the second A entity. 

Yes, second time, but the first time there is no match, and that is where you create second B entity that breaks this. 

> So that begs the question why it breaks when entity B is created before the two A entities.

Because the way it works you are creating A with B together, not B separately.

```ts
    // here you create entity B with value: 'foo'
    const b = new B();
    b.value = 'foo';

    // here you save it to the database
    await orm.em.findOne(b.constructor.name, classToPlain(b)) || await orm.em.persistAndFlush(b);

    // here you create another one
    const b2 = new B();
    b2.value = 'foo';

    // and set it to `a`
    const a = new A();
    a.b = b2;

    console.log(classToPlain(a));
    // here you are trying to search for A, but so far you did not persist any entity A!!
    // so here you are flushing another B entity with the same value: 'foo'
    await orm.em.findOne(a.constructor.name, classToPlain(a)) || await orm.em.persistAndFlush(a);

    // rest of the code is not even reached
    const b3 = new B();
    b3.value = 'foo';

    const a2 = new A();
    a2.b = b3;

    await orm.em.findOne(a2.constructor.name, classToPlain(a2)) || await orm.em.persistAndFlush(a2);
```
